framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java

```diff
@@ -2061,6 +2061,75 @@ protected void adaptGetClassReturnTypeToReceiver(
         classWildcardArg.setExtendsBound(newBound);
     }
 
+    /**
+     * Return the element type of the given iterator. This is usually the type of {@code
+     * expression.itertor().next()}. If {@code expression} is an array, it is the component type of
+     * the array.
+     *
+     * @param expression an expression whose type is an array or implements {@link Iterable}
+     * @return the type of {@code expression.itertor().next()} or if {@code expression} is an array,
+     *     the component type of the array.
+     */
+    public AnnotatedTypeMirror getIteratedType(ExpressionTree expression) {
+        return getIteratedType(expression, getAnnotatedType(expression));
+    }
+
+    /**
+     * Return the element type of the given iterator. This is usually the type of {@code
+     * expression.itertor().next()}. If {@code expression} is an array, it is the component type of
+     * the array.
+     *
+     * @param expression an expression whose type is an array or implements {@link Iterable}
+     * @param iterableType the type of the expression
+     * @return the type of {@code expression.itertor().next()} or if {@code expression} is an array,
+     *     the component type of the array.
+     */
+    protected AnnotatedTypeMirror getIteratedType(
+            ExpressionTree expression, AnnotatedTypeMirror iterableType) {
+        switch (iterableType.getKind()) {
+            case ARRAY:
+                return ((AnnotatedArrayType) iterableType).getComponentType();
+            case WILDCARD:
+                return getIteratedType(
+                        expression,
+                        ((AnnotatedWildcardType) iterableType).getExtendsBound().deepCopy());
+            case TYPEVAR:
+                return getIteratedType(
+                        expression, ((AnnotatedTypeVariable) iterableType).getUpperBound());
+            case DECLARED:
+                TypeElement iterableElement =
+                        ElementUtils.getTypeElement(processingEnv, Iterable.class);
+                AnnotatedDeclaredType iterableElmType = getAnnotatedType(iterableElement);
+                AnnotatedDeclaredType dt =
+                        AnnotatedTypes.asSuper(this, iterableType, iterableElmType);
+                if (dt.getTypeArguments().isEmpty()) {
+                    TypeElement e = ElementUtils.getTypeElement(processingEnv, Object.class);
+                    return getAnnotatedType(e);
+                } else {
+                    return dt.getTypeArguments().get(0);
+                }
+
+                // TODO: Properly desugar Iterator.next(). The below doesn't work because
```

Comment:
Please add a test case and reference that test case from this comment.  Or, if the comment is obviated by other changes such as the addition of the `MethodInvocationNode.iteratorExpression` field, delete the comment.

Reply:
This code is only need if an annotated JDK has annotations on Iterator#next. None of our current checkers have such an annotation, so I can't added a test case for any of the current checkers.  Do you want me to add a test checker that has such an annotation?

Reply:
Thanks for the clarification.  I think it's enough to add a comment explaining what you said here:

> This code is only need if an annotated JDK has annotations on Iterator#next.